### ISSUE TITLE
Document start (---) in yaml inventory config is loaded as a host

### ISSUE THREAD

**Issue by philipsd6** [ansible/ansible#56241]
##### SUMMARY
When using a inventory file for a plugin that loads the inventory as a yaml file (e.g. scaleway.py) if the inventory file has a typical YAML document start (---) as the first line, that document start line is added as a host:
```TASK [setup] **************************************************************************************************************************
fatal: [---]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: unknown option -- -\r\nusage: ssh [-124
6AaCfGaSdFNnqsTtVvXxYy] [-b bind_address] [-c cipher_spec]\n           [-D [bind_address:]port] [-E log_file] [-e escape_char]\n           [-F configfile] [-I pkcs11] [-i identity_file]\n           [-J [user@]host[:port]] [-L address] [-l login_name] [-m mac_spec]\n           [-O ctl_cmd] [-o option] [-p port] [-Q query_option] [-R address]\n           [-S ctl_path] [-W host:port] [-w local_tun[:remote
_tun]]\n           [user@]hostname [command]", "unreachable": true}
ok: [test0000]
ok: [test0002]
ok: [test0001]
```
If the document start line is not in the inventory yaml file, then yaml lint produces this warning message:
```
    /home/.../inventory/my_sample.yml                                                   
      1:1       warning  missing document start "---"  (document-start)                                                               
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
AnsibleLoader (from_yaml())

##### ANSIBLE VERSION

```paste below
ansible 2.7.10
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/philipsd6/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/philipsd6/.local/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Mar 26 2019, 22:13:06) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]

```

##### CONFIGURATION

```paste below
[DEPRECATION WARNING]: DEFAULT_SUDO_USER option, In favor of Ansible Become, which is a generic framework. See become_user. , use 
become instead. This feature will be removed in version 2.8. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
DEFAULT_ASK_PASS(/etc/ansible/ansible.cfg) = True
DEFAULT_INVENTORY_PLUGIN_PATH(/etc/ansible/ansible.cfg) = [u'/usr/share/ansible/plugins/inventory']
DEFAULT_SUDO_USER(/etc/ansible/ansible.cfg) = root
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
INVENTORY_ENABLED(/etc/ansible/ansible.cfg) = [u'host_list', u'script', u'yaml', u'ini', u'auto']

```

##### OS / ENVIRONMENT

Red Hat Enterprise Linux Server release 7.6 (Maipo)

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by bcoca** [comment]
so `---` is optional, I consider this yamllint warning as something to ignore.

That said, the problem here is that the file is being consumed by the ini plugin (the yaml plugin would not use --- as a host). Run the inventory with -vvv and post the output to see the actual errors.

needs_info

**Comment by philipsd6** [comment]
OK, I've finally figured out what's happening. This occurs due to the order of plugins listed in `ansible.cfg`. The default has 'ini' listed before 'auto', so 'ini' tries first, and processes '---' as an actual hostname for some reason, then it passes it on to my custom plugin that adds the actual contents properly.

I consider this a bug, as any inventory plugin should fail without modifying the inventory if the config file isn't valid for that plugin, and the ini plugin does fail:
 ```
[WARNING]:  * Failed to parse /home/philipsd6/devel/project1/inventory/sample.dummy.yml with ini plugin:
/home/philipsd6/devel/project1/inventory/dummy.yml:2: Expected key=value host variable assignment, got:   
dummy
```
...but it still adds '---' to the inventory. If /my/ custom inventory plugin fails, whatever changes it made to the inventory are lost.

As for why this is a problem:
1. Valid YAML should *not* cause errors
2. Yamllint cannot ignore the document-start rule. It either enforces it to be present or absent.
3. molecule uses yamllint and if there's even a warning, the lint stage fails.

Workarounds are:
1. Put all the inventory files in a directory, and then tell yamllint to ignore that directory
2. Put `# yamllint disable-line rule:document-start` at the top of each inventory file
3. Don't use a .yml/.yaml extension on the inventory files.

Solution 1 is not good because ideally, we'd like the sometimes complex inventory plugin config files to be linted.
Solution 2 is just kind of ugly
Solution 3 is not good for the same reason as Solution 1, plus yaml files should have yaml extensions.

**Comment by bcoca** [comment]
So I added `---` detection to ini plugin in #45499, that should be out in the upcoming 2.8 release, we also moved the auto plugin to have higher priority.

The 'partial updated inventory' is a known bug and we are working on it, there are several major hurdles we need to deal with to get that fixed though and will require a redesign of the underlying data store. We hope to have this solved in 2.9 but I don't know if it is something we will be able backport due to the 'expected' intrusive nature of the change.
